### ISSUE TITLE
Week 5 Discussion

### ISSUE THREAD

**Issue by kayvee** [stefanbund/cis155#12]
Hi everyone,

Starting the discussion again (I haven't seen the posted question for the week on Seaport yet).

Hope everyone is off to a good end with the completion of Phase 1! I know that I'm still having to wrap my head around the transaction-portfolio in the context of the buyer.

**Comment by stefanbund** [comment]
Yes, let's begin discussion on that. And let's make our discussions surrounding any coding challenges you have left. Let's talk about the folio. I see it as an array list of transaction objects… What should I talk about?

From Stefan

**Comment by forrestbennett** [comment]
AHHHHHH!!! What happened to my stock symbols? Not sure what happened but somewhere along the way they stopped returning. I'm not feeling too confident in my ability to finish this project this evening. I have been going line by line comparing Prof. code to mine and cant seem to find where it is dropping out. Anyone else come across this issue?
![screenshot 3](https://cloud.githubusercontent.com/assets/17039846/13560454/169186c8-e3d4-11e5-8caa-13715091ad0b.png)

**Comment by stefanbund** [comment]
For your variable market, inside of print market, is that an array list of types stock?

From Stefan

**Comment by forrestbennett** [comment]
I apologize, im not sure what you mean

**Comment by WesleyTo** [comment]
I took a bit of a different approach to the portfolio. I stored transactions in an ArrayList, but I also added a separate data structure for the portfolio (a `HashMap<Stock, Integer>`) that stores a stock and its quantity owned. Then I just wrap up any transactions in a method that will maintain both data structures. The reason was that maintaining a HashMap allows you to check if you own the stock and get the quantity owned in _O(1)_ time. With just an ArrayList, these two operations would take _O(n)_ time in the number of transactions. Of course, the drawback is that this HashMap takes up _O(n)_ memory in the number of stocks owned. It was worth the space trade-off for the speed-up when I had a large number of buyers.

I also did things a bit differently with the price increases. I tied the price change function with a function that handles trades. The trade method will take in a stock and a quantity and programatically change the price of the stock proportional to the quotient of the trade quantity and the IPO quantity. This way, the price can have a positive trajectory when it's being purchased a lot, and a negative trajectory when sold a lot.

I had some trouble dealing with casting/precision in the prices and budgets, but eventually worked it out. I also had a negative infinity error resulting from a division in the price increase. I had to deal with null pointers as a result of the this (I was using `!obj.equals(null)` instead of `obj != null` which threw a `nullPointerException`). I had to set breakpoints and debug to find the problem. I solved it by not changing the ipoQty and maintaining a separate variable to track the number of shares sold. This ended up working out anyway since I needed `ipoQty` to remain constant in my price change calculations.

Problem:

``` java
// Stock.java
private void changePrice(int qty){
    double quotient = ((double)qty) / ipoQty;
    // ipoQty was being decremented elsewhere after a purchase
    //     leading to a zero division
```

Soliution:

``` java
private int qtySold;
...
// change qtySold after a transaction instead of ipoQty
...
public int sharesRemaining(){
    return ipoQty - qtySold;
}
```

**Comment by WesleyTo** [comment]
@forrestbennett
Have you checked your `setSymbol()` method (or whatever you called it) in `Stock.java`. It seems like the stock symbols are being set to empty strings.

**Comment by kayvee** [comment]
@forrestbennett 

Can you include the code for your Stock? That will help us see where the issue is. Right now, all we have is a print statement.
